src/ORM/RelatedData/StandardRelatedDataService.php

```diff
@@ -0,0 +1,507 @@
+<?php
+
+namespace SilverStripe\ORM\RelatedData;
+
+use ReflectionException;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Config\Config;
+use SilverStripe\ORM\DB;
+use SilverStripe\ORM\SS_List;
+use SilverStripe\ORM\ArrayList;
+use SilverStripe\ORM\Connect\Query;
+use SilverStripe\ORM\DataObject;
+use SilverStripe\ORM\DataObjectSchema;
+use SilverStripe\ORM\Queries\SQLSelect;
+
+/**
+ * Service class used to find all other DataObject instances that are related to a DataObject instance
+ * in the database
+ *
+ * Example demonstrating what '$component' and '$componentClassName' variables refer to:
+ * PHP model: private static $has_one = [ 'MyFile' => File::class ]
+ * - $component: 'MyFile'
+ * - $componentClassName: SilverStripe\Assets\File::class
+ *
+ * @internal
+ */
+class StandardRelatedDataService implements RelatedDataService
+{
+
+    /**
+     * Used to prevent duplicate database queries
+     *
+     * @var array
+     */
+    private $queryIdens = [];
+
+    /**
+     * @var array
+     */
+    private $config;
+
+    /**
+     * @var DataObjectSchema
+     */
+    private $dataObjectSchema;
+
+    /**
+     * @var array
+     */
+    private $classToTableName;
+
+    /**
+     * Find all DataObject instances that have a linked relationship with $record
+     *
+     * @param DataObject $record
+     * @param string[] $excludedClasses
+     * @return SS_List
+     */
+    public function findAll(DataObject $record, array $excludedClasses = []): SS_List
+    {
+        // Do not query unsaved DataObjects
+        if (!$record->exists()) {
+            return ArrayList::create();
+        }
+
+        $this->config = Config::inst()->getAll();
+        $this->dataObjectSchema = DataObjectSchema::create();
```

Comment:
You should be using ::singleton() not ::create().

Your issue with test classes is that sapphiretest will flush the cache of dataobject singleton, not ALL instances that are created. :)

Or just DataObject::getSchema() instead of storing it in a protected property.